### PR TITLE
Fixing context elastic

### PR DESCRIPTION
**Pull request by jrmbrgs**

### CODE REVIEW

File: Library/ContextStorageAdapters/ElasticsearchAdapter.php [hunk 1]
```diff
@@ -95,7 +98,8 @@ public function initialize(Json $config)
      */
     private function checkVendorLibraryAvailable($className)
     {
-        if (!class_exists($className)) {
+        //$f = new $className;
```
Comment (adefretin) [inline Library/ContextStorageAdapters/ElasticsearchAdapter.php]: to remove

File: Library/ContextStorageAdapters/ElasticsearchAdapter.php [hunk 2]
```diff
@@ -95,7 +98,8 @@ public function initialize(Json $config)
...
+        if (!class_exists($className) && false) {
```
Comment (adefretin) [inline Library/ContextStorageAdapters/ElasticsearchAdapter.php]: remove && false

File: Library/ContextStorageAdapters/ElasticsearchAdapter.php [hunk 3]
```diff
@@ -197,14 +202,16 @@ private function initClient()
...
         $this->checkParameters([$workflowProcessId]);
 
         try {
             $requestParamHash = array_merge(
                 ['id' => $workflowProcessId],
                 $this->commonRequestParamHash
             );
-            return $this->client->get($requestParamHash);
+            $resultHash = $this->client->get($requestParamHash);
+            return $resultHash['_source'];
```
Comment (adefretin) [inline Library/ContextStorageAdapters/ElasticsearchAdapter.php]: perhpas add keyword "_source" as const   DOC_SOURCE

File: Library/ContextStorageAdapters/ElasticsearchAdapter.php [hunk 4]
```diff
@@ -284,6 +294,7 @@ public function save(string $workflowProcessId, array $documentHash) : array
                 return $this->client->index(array_merge($requestParamHash, ['body' => $documentHash]));
             }
         } catch (\Exception $exception) {
+            echo $exception->getMessage();
```
Comment (adefretin) [inline Library/ContextStorageAdapters/ElasticsearchAdapter.php]: to remove

File: Library/Services/ContextStorage.php [hunk 5]
```diff
@@ -227,10 +241,52 @@ public function getWorkflowStepList(string $workflowProcessId) : array {
...
+    /**
+     * Update the given step related to the given workflow/stepcode
+     *
+     * @param string $workflowProcessId the workflow identifier
+     * @param string $stepCode          the step code
+     * @param array  $stepHash          the step data to save
+     *
+     * @returns array
+     */
+    public function updateWorkflowStep(string $workflowProcessId, $stepCode, $stepHash)
```
Comment (adefretin) [inline Library/Services/ContextStorage.php]: missing param type

File: Library/Services/ContextStorage.php [hunk 6]
```diff
@@ -227,10 +241,52 @@ public function getWorkflowStepList(string $workflowProcessId) : array {
...
+    {
+        $this->di->get('logger')->debug(json_encode(func_get_args()));
+        $script = <<<EOT
+        def nbi=ctx._source.workflow.steps.size();
```
Comment (adefretin) [inline Library/Services/ContextStorage.php]: naming nbi => nbStep

File: Library/Services/ContextStorage.php [hunk 7]
```diff
@@ -227,10 +241,52 @@ public function getWorkflowStepList(string $workflowProcessId) : array {
...
+        for (i=0; i < nbi; i++) {
```
Comment (adefretin) [inline Library/Services/ContextStorage.php]: naming i => stepIndex

File: Library/Services/ContextStorage.php [hunk 8]
```diff
@@ -227,10 +241,52 @@ public function getWorkflowStepList(string $workflowProcessId) : array {
...
+            if (ctx._source.workflow.steps[i] instanceof List) {
+                def nbj=ctx._source.workflow.steps[i].size();
```
Comment (adefretin) [inline Library/Services/ContextStorage.php]: naming nbj => nbParallelizedStep

File: Library/Services/ContextStorage.php [hunk 9]
```diff
@@ -227,10 +241,52 @@ public function getWorkflowStepList(string $workflowProcessId) : array {
...
+                for (j=0; j < nbj; j++) {
```
Comment (adefretin) [inline Library/Services/ContextStorage.php]: naming j => parallelizedStepIndex

File: Library/Services/ContextStorage.php [hunk 10]
```diff
@@ -227,10 +241,52 @@ public function getWorkflowStepList(string $workflowProcessId) : array {
...
+                        ctx._source.workflow.steps[i][j] = stepHash;
+                        break;
+                    }
+                }
+            } else if (ctx._source.workflow.steps[i].name == stepCode) {
+                ctx._source.workflow.steps[i] =stepHash;
+            }
+        }
+EOT;
+        $updateHash = [
```
Comment (adefretin) [inline Library/Services/ContextStorage.php]: we can add updatePartial  on elasticsearchadapter.

updatePartial(string $workflowProcessId, string updatePartialName, array $paramHash)

switch (updatePartialName)
  case 'updateWorkflowStep':
            $script = <<<EOT....
this->adapter->update($workflowProcessId, $updateHash);

File: Library/Services/ContextStorage.php [hunk 11]
```diff
@@ -227,10 +241,52 @@ public function getWorkflowStepList(string $workflowProcessId) : array {
...
+     * @param string $workflowProcessId the workflow identifier
+     * @param string $stepCode          the step code
+     * @param array  $stepHash          the step data to save
+     *
+     * @returns array
+     */
+    public function updateWorkflowStep(string $workflowProcessId, $stepCode, $stepHash)
+    {
+        $this->di->get('logger')->debug(json_encode(func_get_args()));
+        $script = <<<EOT
```
Comment (adefretin) [inline Library/Services/ContextStorage.php]: perhaps we can check if context storage adapter manage partial update.

if ($this->adpater->hasUpdatePartial())
   // use update partial. ..
else 
   // use update classic

File: Library/Services/ContextStorage.php (hunk 10 above)
Comment (jrmbrgs) [reply Library/Services/ContextStorage.php]: not proud too about my first implementation ... 
We obviously have to keep the `ContextStorage` not related to any storage engine.
I did not found a great solution yet ...

File: Library/Services/ContextStorage.php (hunk 11 above)
Comment (jrmbrgs) [reply Library/Services/ContextStorage.php]: partial update is required, no choice.

**Review by adefretin — APPROVED** [review]

**Pull request merged**